File: src/main/java/org/apache/prepbuddy/smoothers/SlidingWindow.java
package org.apache.prepbuddy.smoothers;

import java.io.Serializable;
import java.util.LinkedList;

/**
 * A window for smoothing data.
 */
public abstract class SlidingWindow implements Serializable {
    protected int size;
    protected LinkedList<Double> queue;

    public SlidingWindow(int size) {
        this.size = size;
        queue = new LinkedList<>();
    }

    public void add(double value) {
        if (isFull())
            queue.remove();
        queue.add(value);
    }

    public boolean isFull() {
        return queue.size() == size;
    }

    public Double sum() {
        Double sum = 0.0;
        for (Double oneValue : queue) {
            sum += oneValue;
        }
        return sum;
    }

    public abstract Double average();
}

File: src/main/scala/org/apache/datacommons/prepbuddy/smoothers/SlidingWindow.scala
package org.apache.datacommons.prepbuddy.smoothers

import scala.collection.mutable

abstract class SlidingWindow(windowSize: Int) extends Serializable {
    protected var windowValues: mutable.Queue[Double] = mutable.Queue()

    def add(value: Double) {
        if (isFull) windowValues.dequeue()
        windowValues += value
    }

    def isFull: Boolean = windowValues.length == windowSize

    def average: Double
}
